File: src/algorithm/sorting/insertion-sort/insertionSort.js
#! /usr/local/bin/node


class InsertionSort {
    sort(originalList=[]) {
        let list = [...originalList];
        if (list.length <= 1) {
            return list;
        }

        for (let j = 1; j < list.length; j++) {
            let current = list[j];
            let i = j - 1;
            while (i >= 0) {
                if (current >= list[i]) {
                    list[i + 1] = current;
                    break;
                }
                list[i + 1] = list[i];
                list[i] = current;
                i--;
            }
        }

        return list;
    }

    isSorted(array=[]) {
        if (array.length <= 1) {
            return true;
        }

        for (let i = 1; i < array.length; i++) {
            if (array[i] < array[i - 1]) {
                return false;
            }
        }

        return true;
    }

    test() {
        const array = new Array(40000).fill(undefined).map(item => Math.random() * 10000);
        console.time('sort');
        let result = this.sort(array);
        console.timeEnd('sort');
        let sorted = this.isSorted(result);
        console.log('sorted:', sorted);
    }
}

new InsertionSort().test();

File: src/algorithm/sorting/insertion-sort/insertion_sort.py
#! /usr/local/bin/python3


class InsertionSort:
    def sort(self, original_list):
        if original_list is None:
            return []

        arr = original_list.copy()

        if len(arr) <= 1:
            return arr

        sorted = [arr.pop()]

        while len(arr) > 0:
            item = arr.pop()

            for i, val in enumerate(sorted):
                if val >= item:
                    sorted = sorted[0:i] + [item] + sorted[i:len(sorted)]
                    break
            else:
                sorted.append(item)

        return sorted

    def test(self):
        import time

        start = time.process_time()
        result = self.sort([23, 2, 6, 3, 77, 4, 98, 8, 5, 5, 8])
        end = time.process_time()
        print(result, (end - start) * 1000)

if __name__ == '__main__':
    InsertionSort().test()
